{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b6eb7bfa-226b-49fd-b860-92c2c94c4e40",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import multiprocessing\n",
    "pd.options.display.max_columns = None\n",
    "pd.options.display.max_rows = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1196a041-3c89-47cb-b68b-d2e1845cf06f",
   "metadata": {},
   "outputs": [],
   "source": [
    "binsize=25 #fix number of CpG\n",
    "maxgap=500 #max length of gap between bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c81f8204-fbe5-4b10-b5aa-b3863386714e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # all CpGs in the human genome\n",
    "# df=pd.read_csv(\"/t2/ny/ref/hg38_autoxy_cg.bed\",sep=\"\\t\",header=None,dtype={0:str},usecols=[0,1,2,5])\n",
    "# df.columns=[\"chro\",\"start\",\"end\",\"strand\"]\n",
    "\n",
    "# # binning\n",
    "# nbin = 1\n",
    "# bin_column = [1]\n",
    "# chromStart = df['start'].values\n",
    "# chrom = df['chro'].values\n",
    "\n",
    "# present_size=1\n",
    "# for i in range(1, df.shape[0]):\n",
    "#     present_size+=1\n",
    "#     if chromStart[i] - chromStart[i-1] > maxgap or chrom[i] != chrom[i-1] or present_size>binsize*2:\n",
    "#         nbin += 1\n",
    "#         present_size=1\n",
    "#     bin_column.append(nbin)\n",
    "# df['bin'] = bin_column\n",
    "\n",
    "# # remove bins that do not reach binsize\n",
    "# df_only_in_binsize = df.groupby('bin').filter(lambda x: len(x) == binsize*2)\n",
    "# total_bin_num=len(df_only_in_binsize.bin.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edb3fc0c-51ec-4846-9f4a-3815d5da2cac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2234ac67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Directly use the ready-made binned bed\n",
    "df_only_in_binsize=pd.read_csv(\"/t4/like/aml/bin25CpG_500gap.bed\",sep=\"\\t\",header=None,dtype={0:str})\n",
    "df_only_in_binsize.columns=[\"chro\",\"start\",\"end\",\"bin\",'strand']\n",
    "total_bin_num=len(df_only_in_binsize.bin.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "78456edd-4b3f-44e9-b2db-97ddb40a693d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def vectorized_similarity(data):\n",
    "    nan_mask = ~np.isnan(data)\n",
    "    num_rows = data.shape[0]\n",
    "    \n",
    "    data_expanded_1 = np.expand_dims(data, axis=1)\n",
    "    data_expanded_2 = np.expand_dims(data, axis=0)\n",
    "    mask_expanded_1 = np.expand_dims(nan_mask, axis=1)\n",
    "    mask_expanded_2 = np.expand_dims(nan_mask, axis=0)\n",
    "\n",
    "    valid_mask = mask_expanded_1 & mask_expanded_2\n",
    "    equality = (data_expanded_1 == data_expanded_2) & valid_mask\n",
    "\n",
    "    valid_counts = np.sum(valid_mask, axis=2)\n",
    "    equal_counts = np.sum(equality, axis=2)\n",
    "\n",
    "    valid_counts[valid_counts == 0] = 1\n",
    "    similarity = equal_counts / valid_counts\n",
    "\n",
    "    return similarity,valid_counts\n",
    "    \n",
    "def get_read_m_array(read,l_cg_pos):\n",
    "    d_m_bases=read.modified_bases\n",
    "    d_methy=dict(zip(l_cg_pos,[np.nan]*len(l_cg_pos)))\n",
    "    ref_positions = read.get_reference_positions(full_length=True)\n",
    "    if ('C', 0, 'm') in d_m_bases:\n",
    "        strand_flag=0\n",
    "    elif ('C', 1, 'm') in d_m_bases:\n",
    "        strand_flag=1\n",
    "    elif d_m_bases=={}:\n",
    "        return [False,False]\n",
    "    else:\n",
    "        raise Exception(f\"cant handle {d_m_bases} in {chrom}:{start_pos}-{end_pos}\")\n",
    "\n",
    "    # Only take the pos of the corresponding strand\n",
    "    l_cg_pos=l_cg_pos[strand_flag::2]\n",
    "    \n",
    "    d_methy=dict(zip(l_cg_pos,[np.nan]*len(l_cg_pos)))\n",
    "    for i in d_m_bases[('C', strand_flag, 'm')]:\n",
    "        pos=ref_positions[i[0]]\n",
    "        if pos in l_cg_pos:\n",
    "            if i[1]>204:\n",
    "                d_methy[pos]=1\n",
    "            elif i[1]<51:\n",
    "                d_methy[pos]=0\n",
    "    l_methy=list(d_methy.values())\n",
    "    return l_methy,strand_flag\n",
    "\n",
    "def get_read_matrix(dfbin,bam,l_cg_pos,strand='both'):\n",
    "    \n",
    "    start_pos=l_cg_pos[0]\n",
    "    end_pos=l_cg_pos[-1]\n",
    "    chrom=dfbin['chro'].iloc[0]\n",
    "    reads = bam.fetch(chrom, start_pos, end_pos)\n",
    "\n",
    "    read_matrix=[]\n",
    "\n",
    "    for read in reads: \n",
    "        if read.is_unmapped or read.is_secondary or read.is_supplementary:\n",
    "            continue\n",
    "        if read.has_tag(\"MM\"):\n",
    "            l_methy,strand_flag=get_read_m_array(read,l_cg_pos)            \n",
    "            if not l_methy:\n",
    "                continue\n",
    "            if strand!='both' and strand_flag!=strand:\n",
    "                continue\n",
    "            read_matrix.append(l_methy)\n",
    "                \n",
    "    read_matrix=np.array(read_matrix)\n",
    "    return read_matrix\n",
    "\n",
    "def get_similarity_summary(dfbin,bam,l_cg_pos,strand='both'):\n",
    "    read_matrix=get_read_matrix(dfbin,bam,l_cg_pos,strand=strand)\n",
    "    if np.size(read_matrix):\n",
    "        similarity_matrix,valid_counts=vectorized_similarity(read_matrix)\n",
    "        half_diagonal_similarity_matrix = similarity_matrix[np.triu_indices(len(similarity_matrix), k=1)]\n",
    "        if half_diagonal_similarity_matrix.size > 0:\n",
    "            mean_value = np.nanmean(half_diagonal_similarity_matrix)\n",
    "        else:\n",
    "            mean_value = np.nan\n",
    "            \n",
    "        half_diagonal_valid_counts = valid_counts[np.triu_indices(len(valid_counts), k=1)]\n",
    "        total_overlap_point=np.sum(half_diagonal_valid_counts)\n",
    "        \n",
    "        reads=len(similarity_matrix)\n",
    "        overlap_read_count = np.count_nonzero(~np.isnan(half_diagonal_similarity_matrix))\n",
    "    else:\n",
    "        reads=overlap_read_count=total_overlap_point=0\n",
    "        mean_value=np.nan\n",
    "    return mean_value,overlap_read_count,reads,total_overlap_point\n",
    "\n",
    "def get_mpercentage(dfbin,bam,l_cg_pos):\n",
    "    read_matrix=get_read_matrix(dfbin,bam,l_cg_pos)\n",
    "    if np.size(read_matrix):\n",
    "        read_matrix=get_read_matrix(dfbin,bam,l_cg_pos)\n",
    "        m=np.sum(read_matrix == 1)\n",
    "        unm=np.sum(read_matrix == 0)\n",
    "        if m+unm==0:\n",
    "            return np.nan\n",
    "        return m/(m+unm)\n",
    "    return np.nan\n",
    "    \n",
    "    \n",
    "def get_summary(dfbin,bamfile):\n",
    "    bam = pysam.AlignmentFile(bamfile, 'rb')\n",
    "    l_cg_pos=list(dfbin['start'])\n",
    "    dfbin_0,dfbin_last=dfbin.index[0],dfbin.index[-1]\n",
    "    start=dfbin['start'][dfbin_0]\n",
    "    end=dfbin['end'][dfbin_last]\n",
    "    mean_value,overlap_read_count,reads,total_overlap_point=get_similarity_summary(dfbin,bam,l_cg_pos,strand='both')\n",
    "#     f_mean_value,f_overlap_read_count,f_reads,f_total_overlap_point=get_similarity_summary(dfbin,bam,l_cg_pos,strand=0)\n",
    "#     r_mean_value,r_overlap_read_count,r_reads,r_total_overlap_point=get_similarity_summary(dfbin,bam,l_cg_pos,strand=1)\n",
    "#     mpercentage=get_mpercentage(dfbin,bam,l_cg_pos)\n",
    "    summary={\n",
    "    'chr':dfbin['chro'].iloc[0],\n",
    "    'start':start,\n",
    "    'end':end,\n",
    "    'length':end-start,\n",
    "#     'mpercentage':mpercentage,\n",
    "        \n",
    "    'similarity_mean':mean_value,\n",
    "    'n_overlapped_reads_pairs':overlap_read_count,\n",
    "    'n_reads':reads,\n",
    "    'total_overlapped_points':total_overlap_point,\n",
    "        \n",
    "#     'f_similarity_mean':f_mean_value,\n",
    "#     'f_n_overlapped_reads_pairs':f_overlap_read_count,\n",
    "#     'f_n_reads':f_reads,\n",
    "#     'f_total_overlapped_points':f_total_overlap_point,\n",
    "        \n",
    "#     'r_similarity_mean':r_mean_value,\n",
    "#     'r_n_overlapped_reads_pairs':r_overlap_read_count,\n",
    "#     'r_n_reads':r_reads,\n",
    "#     'r_total_overlapped_points':r_total_overlap_point,\n",
    "    }\n",
    "    \n",
    "    \n",
    "    return summary            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ab752dff-85de-45fb-ad19-7471f0a3466f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "# path='/t2/ny/dac/dec_cel_bammod'\n",
    "path='/t2/ny/dnmt/modbam_drd'\n",
    "name_list=[i[:-8] for i in os.listdir(path) if i.endswith('_M_s.bam')]\n",
    "\n",
    "d_sample={}\n",
    "for i in name_list:\n",
    "#     d_sample[i]='/t2/ny/dac/dec_cel_bammod'\n",
    "    d_sample[i]='/t2/ny/dnmt/modbam_drd'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "20ae8ee6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MT1340_M_s.bam          MT2658_M_s.bam      WT561_M_s.bam.bai\r\n",
      "MT1340_M_s.bam.bai      MT2658_M_s.bam.bai  WT621_M_s.bam\r\n",
      "MT2136_5hmcM.bam        MT614_M_s.bam       WT621_M_s.bam.bai\r\n",
      "MT2136_5hmcM_s.bam      MT614_M_s.bam.bai   WT730_M_s.bam\r\n",
      "MT2136_5hmcM_s.bam.bai  MT615_M_s.bam       WT730_M_s.bam.bai\r\n",
      "MT2136_M_s.bam          MT615_M_s.bam.bai   WT881_M_s.bam\r\n",
      "MT2136_M_s.bam.bai      WT1027_M_s.bam      WT881_M_s.bam.bai\r\n",
      "MT2137_M_s.bam          WT1027_M_s.bam.bai  WT898_M_s.bam\r\n",
      "MT2137_M_s.bam.bai      WT2795_M_s.bam      WT898_M_s.bam.bai\r\n",
      "MT2652_M_s.bam          WT2795_M_s.bam.bai\r\n",
      "MT2652_M_s.bam.bai      WT561_M_s.bam\r\n"
     ]
    }
   ],
   "source": [
    "ls /t2/ny/dnmt/modbam_drd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "70bf70fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import os\n",
    "# path='/t2/ny/dac/dec_cel_modbam_drdsup'\n",
    "# name_list=[i[:-8] for i in os.listdir(path) if i.endswith('K052_DMSO_M_s.bam')]\n",
    "\n",
    "# d_sample={}\n",
    "# for i in name_list:\n",
    "#     d_sample[i]=path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d42a2129",
   "metadata": {},
   "outputs": [],
   "source": [
    "# d_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5c258214-46a6-4b03-8169-e525c72cc05c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# nonres=['LK_2671_D10','CWH_2713_D10','WCMA_2756_D0','CWH_2713_D0','HMS_2075_D0','LK_2671_D0','LCK_1341_D0','HMS_2075_D10','LCK_1341_D10','WCMA_2756_D10']\n",
    "# res=['NKW_2685_D10', 'TTK_2749_D0', 'LRL_2827_D0', 'LCK_2643_D10', 'MYK_3054_D0', 'CBM_3053_D10', 'NKW_2685_D0', 'LCK_2643_D0', 'CBM_3053_D0', 'LWK_2966_D0', 'WLM_3033_D10', 'TTK_2749_D10', 'CWH_3113_D10', 'LLC_2861_D0', 'CWFS_2816_D0', 'LLC_2861_D10', 'FYC_3026_D10', 'LWK_2966_D10', 'CWH_3113_D0', 'FYC_3026_D0', 'LRL_2827_D10', 'WLM_3033_D0', 'CWFS_2816_D10', 'SKK_2719_D0', 'SKK_2719_D10', 'MYK_3054_D10']\n",
    "# pbsc=['PBSC1', 'PBSC3', 'PBSC2']\n",
    "\n",
    "# d_sample={}\n",
    "# for i in nonres:\n",
    "#     d_sample[i]='/t2/ny/dac/modbamall_drd/nonres'\n",
    "# for i in res:\n",
    "#     d_sample[i]='/t2/ny/dac/modbamall_drd/res'\n",
    "# for i in pbsc:\n",
    "#     d_sample[i]='/t2/ny/dac/modbamall_drd/pbsc'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "54dc08ad-e835-4f85-8f57-c0598233fdb1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cef609d5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66435aec-7197-416c-9268-27605b33019d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample_name in name_list:\n",
    "# for sample_name in nonres+res+pbsc:\n",
    "\n",
    "    bamfile=f'{d_sample[sample_name]}/{sample_name}_M_s.bam'\n",
    "    pool = multiprocessing.Pool(processes=128)\n",
    "    l_summary = []\n",
    "    for i in range(total_bin_num):\n",
    "        result = pool.apply_async(get_summary, (df_only_in_binsize[i*binsize*2:(i+1)*binsize*2], bamfile))\n",
    "        l_summary.append(result)\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "    \n",
    "    l_summary = [res.get() for res in l_summary]\n",
    "    df_summary = pd.DataFrame(l_summary)\n",
    "    \n",
    "    df_summary.to_csv(f'bin{binsize}CpG_{maxgap}gap_{sample_name}.csv',index=False,encoding='utf_8_sig')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4505291f-5b29-4785-9fa1-ea581e9ce98c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# example\n",
    "\n",
    "\n",
    "sample_name='LK_2671_D10'\n",
    "bamfile=f'/t2/ny/dac/modbamall_drd/nonres/{sample_name}_M_s.bam'\n",
    "# sample_name=name_list[0]\n",
    "# bamfile=f'{d_sample[sample_name]}/{sample_name}_s.bam'\n",
    "\n",
    "i=45\n",
    "dfbin=df_only_in_binsize[i*binsize*2:(i+1)*binsize*2]\n",
    "\n",
    "bam = pysam.AlignmentFile(bamfile, 'rb')\n",
    "\n",
    "l_cg_pos=list(dfbin['start'])\n",
    "\n",
    "start_pos=l_cg_pos[0]\n",
    "end_pos=l_cg_pos[-1]\n",
    "chrom=dfbin['chro'].iloc[0]\n",
    "reads = bam.fetch(chrom, start_pos, end_pos)\n",
    "\n",
    "read_matrix=[]\n",
    "\n",
    "for read in reads: \n",
    "    if read.is_unmapped or read.is_secondary or read.is_supplementary:\n",
    "        continue\n",
    "    if read.has_tag(\"MM\"):\n",
    "        l_methy,strand_flag=get_read_m_array(read,l_cg_pos)            \n",
    "        if not l_methy:\n",
    "            continue\n",
    "        read_matrix.append(l_methy)\n",
    "\n",
    "read_matrix=np.array(read_matrix)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4f0d113b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chro</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>bin</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2250</th>\n",
       "      <td>1</td>\n",
       "      <td>134236</td>\n",
       "      <td>134237</td>\n",
       "      <td>80</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2251</th>\n",
       "      <td>1</td>\n",
       "      <td>134237</td>\n",
       "      <td>134238</td>\n",
       "      <td>80</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2252</th>\n",
       "      <td>1</td>\n",
       "      <td>134264</td>\n",
       "      <td>134265</td>\n",
       "      <td>80</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2253</th>\n",
       "      <td>1</td>\n",
       "      <td>134265</td>\n",
       "      <td>134266</td>\n",
       "      <td>80</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2254</th>\n",
       "      <td>1</td>\n",
       "      <td>134295</td>\n",
       "      <td>134296</td>\n",
       "      <td>80</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chro   start     end  bin strand\n",
       "2250    1  134236  134237   80      +\n",
       "2251    1  134237  134238   80      -\n",
       "2252    1  134264  134265   80      +\n",
       "2253    1  134265  134266   80      -\n",
       "2254    1  134295  134296   80      +"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfbin.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
